File: src/sar/viz/drifters.py
"""
drifters.py -- publish the GDP drifter observations for the site.

    python -m sar.viz.drifters --data /home/26p67/data --out /home/26p67/data/published

The third dataset, and the one that is neither a gridded field nor engine
output. It is the **validation** set: R2's hindcast compares simulated drift
against these real trajectories, and the `track` layer is how that comparison
becomes a picture rather than a table of errors.

PARQUET, NOT ZARR. Zarr is for regular gridded arrays and these are nothing of
the kind -- irregular positions, irregular times, one row per observation per
buoy. Parquet is the format for that, it is what Hugging Face's dataset viewer
reads, and `pyarrow` is already pinned. Forcing a trajectory into a grid would
throw away the thing that makes it a trajectory.

WHAT IS PUBLISHED, AND WHY IT IS NOT THE RAW CSV. `load_drifters` applies every
rule `sar.fetch.drifters.report()` established -- axis renames, longitude to
0-360, SST Kelvin to Celsius masked against fill values, `undrogued` per
OBSERVATION rather than per buoy, and `segment_id` split on gaps over 3 h. The
raw CSV is 103.8 MB of ERDDAP's conventions with a units row wedged under the
header; republishing that would hand everyone else the same four traps.

Longitude is converted to -180..180 here, as in `sar.viz.archive`: this is the
presentation boundary and a web map cannot use 0-360.
"""

import argparse
import json
from datetime import datetime, timezone
from pathlib import Path

import numpy as np
import pandas as pd

from sar.utils.data_io import load_drifters
from sar.utils.geo import to_display_longitude

# Columns worth shipping. `sst_c` earns its place because D013's survivable
# window is a function of water temperature, so the same file answers "where
# did it drift" and "how long would someone last there".
COLUMNS = ["ID", "time", "lat", "lon", "segment_id", "undrogued", "sst_c", "ve", "vn"]

# A segment shorter than this cannot test a 48 h drift forecast, which is what
# D018 budgets the validation runs against.
#
# DURATION, NOT OBSERVATION COUNT, and the difference is not pedantry. 48 hourly
# observations span 47 hours -- the usual fencepost -- and a 47 h segment cannot
# verify a 48 h forecast. Measured 2026-09-18 on the full five-year pull:
#
#     segments total                     825
#     duration      >= 48 h              640   <- what R2 can actually use
#     observations  >= 48                642   <- what D018 cites
#
# Two segments have 48 observations across 47 hours. D018's 642 counts
# observations; for its own stated purpose the number is 640.
MIN_SEGMENT_HOURS = 48


def prepare(df: pd.DataFrame) -> pd.DataFrame:
    """Trim to the publishable columns and convert to display longitude."""
    keep = [c for c in COLUMNS if c in df.columns]
    out = df[keep].copy()
    out["lon"] = to_display_longitude(out["lon"].to_numpy())
    # float32 halves the file and is far finer than the data's own precision:
    # a GDP position is good to ~100 m, which is 1e-3 deg, while float32 holds
    # about 1e-5 deg at these magnitudes.
    for c in ("lat", "lon", "sst_c", "ve", "vn"):
        if c in out:
            out[c] = out[c].astype("float32")
    return out.sort_values(["ID", "time"]).reset_index(drop=True)


def segment_summary(df: pd.DataFrame) -> pd.DataFrame:
    """One row per contiguous in-box run: the unit R2 validates against.

    The site draws tracks from this rather than scanning a million rows to work
    out what a track even is, and it is the table that answers "how many
    independent validation cases do we actually have".
    """
    g = df.groupby("segment_id", sort=True)
    out = pd.DataFrame({
        "segment_id": g.size().index,
        "buoy": g["ID"].first().to_numpy(),
        "start": g["time"].min().to_numpy(),
        "end": g["time"].max().to_numpy(),
        "n": g.size().to_numpy(),
        "lat0": g["lat"].first().to_numpy(),
        "lon0": g["lon"].first().to_numpy(),
        # Per observation, so a buoy that loses its drogue mid-segment is
        # partly undrogued rather than being forced into one camp.
        "undrogued_frac": g["undrogued"].mean().to_numpy(),
    })
    hours = (out["end"] - out["start"]) / np.timedelta64(1, "h")
    out["hours"] = hours.astype("float32")
    return out.sort_values("start").reset_index(drop=True)


def publish(data: Path, out: Path, source: Path | None = None) -> dict:
    if source is None:
        files = sorted((data / "raw").glob("gdp_hourly_*.csv"))
        if not files:
            raise FileNotFoundError(f"no gdp_hourly_*.csv in {data / 'raw'}")
        if len(files) > 1:
            # Refuse rather than guess. The first version took files[-1], which
            # is "last alphabetically" -- against the three local pulls that
            # silently selected the SMALLEST (a 3-month slice) over the full
            # five-year file, and published it without a word. Picking the
            # largest would work today and break the day a bigger partial pull
            # appears, so the caller says which.
            listing = "\n  ".join(
                f"{f.name}  ({f.stat().st_size / 1e6:.1f} MB)" for f in files)
            raise SystemExit(
                f"{len(files)} drifter CSVs in {data / 'raw'} -- pass --source to "
                f"say which:\n  {listing}"
            )
        source = files[0]

    print(f"reading   {source.name} ({source.stat().st_size / 1e6:.1f} MB)")
    obs = prepare(load_drifters(source))
    segs = segment_summary(obs)
    long_enough = segs[segs["hours"] >= MIN_SEGMENT_HOURS]

    out.mkdir(parents=True, exist_ok=True)
    obs_path = out / "drifters.parquet"
    seg_path = out / "drifter_segments.parquet"
    obs.to_parquet(obs_path, index=False, compression="zstd")
    segs.to_parquet(seg_path, index=False, compression="zstd")

    manifest = {
        "generated": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "product": "drifters",
        "label": "NOAA GDP drifters (hourly, quality-controlled)",
        "longitude_convention": "-180..180 (display; D020 stores 0-360)",
        "files": {
            "observations": {"path": obs_path.name, "rows": int(len(obs)),
                             "bytes": obs_path.stat().st_size},
            "segments": {"path": seg_path.name, "rows": int(len(segs)),
                         "bytes": seg_path.stat().st_size},
        },
        "counts": {
            "observations": int(len(obs)),
            "buoys": int(obs["ID"].nunique()),
            "segments": int(len(segs)),
            f"segments_over_{MIN_SEGMENT_HOURS}h_duration": int(len(long_enough)),
            "undrogued_fraction": float(obs["undrogued"].mean()),
        },
        "period": {"start": str(obs["time"].min()), "end": str(obs["time"].max())},
        "bbox": [float(obs["lat"].min()), float(obs["lon"].min()),
                 float(obs["lat"].max()), float(obs["lon"].max())],
        "provenance": {"source_file": source.name, "script": "sar.viz.drifters"},
    }
    (out / "drifters_archive.json").write_text(json.dumps(manifest, indent=2), encoding="utf-8")

    c = manifest["counts"]
    print(f"wrote     {obs_path.name}  {len(obs):,} obs, "
          f"{obs_path.stat().st_size / 1e6:.1f} MB "
          f"(from {source.stat().st_size / 1e6:.0f} MB of CSV)")
    print(f"wrote     {seg_path.name}  {len(segs):,} segments, "
          f"{seg_path.stat().st_size / 1e3:.0f} kB")
    print(f"buoys     {c['buoys']}")
    print(f"undrogued {c['undrogued_fraction'] * 100:.2f} %")
    print(f"segments  {len(segs):,} total, {len(long_enough):,} lasting at least "
          f"{MIN_SEGMENT_HOURS} h (duration, not observation count)")
    print(f"period    {manifest['period']['start'][:10]} -> {manifest['period']['end'][:10]}")
    return manifest


# ---------------------------------------------------------------------------
# The track export the site's drifter layer reads (issue #50).
#
#   drifter_index.json          one row per buoy: small, loaded once
#   drifter_tracks/<ID>.json    one buoy's every fix, fetched when it is needed
#
# SPLIT BY BUOY, NOT ONE FILE, because the site shows one buoy's path at a time
# and a few buoys' positions at once. All of them together are tens of MB; one
# is tens of kB, so a click costs one small fetch and the path keeps its full
# hourly resolution rather than being thinned to fit one download.
# ---------------------------------------------------------------------------

INDEX_NAME = "drifter_index.json"
TRACK_DIR = "drifter_tracks"

# 4 decimal places of a degree is 11 m, far finer than a GDP position is good for
# (~100 m), and it keeps the files small.
COORD_DP = 4

# A fix further than this from a real GPS fix is the fitting method's guess, and
# is drawn dashed (#55: 1.49 % of hourly points).
FAR_FROM_FIX_H = 3.0


def _iso(ts: pd.Timestamp) -> str:
    """UTC ISO 8601 with a Z, which every browser's Date parses the same way."""
    return pd.Timestamp(ts).tz_convert("UTC").strftime("%Y-%m-%dT%H:%M:%SZ")


def _tier_summary(g: pd.DataFrame) -> str:
    """One word for a buoy's drogue record: drogued, undrogued, mixed or uncertain."""
    if g["tier_uncertain"].any():
        return "uncertain"
    und = g["undrogued"]
    if und.all():
        return "undrogued"
    return "mixed" if und.any() else "drogued"


def track_record(g: pd.DataFrame) -> dict:
    """One buoy's fixes as compact columns, in time order.

    `h` is whole hours since `t0` (every fix is on the hour); `seg` renumbers the
    buoy's in-box segments from 0, so the site never draws a line across a gap;
    `und` and `far` are 0/1 per fix. Longitude is the display convention.
    """
    g = g.sort_values("time")
    t0 = g["time"].iloc[0]
    hours = ((g["time"] - t0).dt.total_seconds() / 3600.0).round().astype(int)
    seg = g["segment_id"].rank(method="dense").astype(int) - 1
    far = (g["fix_gap_h"] > FAR_FROM_FIX_H) if "fix_gap_h" in g.columns else pd.Series(False, index=g.index)
    return {
        "id": str(g["ID"].iloc[0]),
        "t0": _iso(t0),
        "h": hours.tolist(),
        "lat": g["lat"].round(COORD_DP).tolist(),
        "lon": np.round(to_display_longitude(g["lon"].to_numpy()), COORD_DP).tolist(),
        "seg": seg.tolist(),
        "und": g["undrogued"].astype(int).tolist(),
        "far": far.fillna(False).astype(int).tolist(),
    }


def buoy_index(df: pd.DataFrame, units: pd.DataFrame | None = None) -> list[dict]:
    """One entry per buoy, sorted by first fix: what the dots, list and search need.

    `splits` lists the validation splits the buoy's units fall in (#51), empty if
    none of its runs is long enough to be a unit. `sealed` is true if any unit is
    sealed or in the 2023 holdout: the raw track may be looked at, but the engine
    must not be compared against it before the frozen evaluation run (D025).
    """
    splits: dict[str, list[str]] = {}
    if units is not None and len(units):
        for bid, s in units.groupby(units["ID"].astype(str))["split"]:
            splits[bid] = sorted(set(s))
    rows = []
    for bid, g in df.groupby(df["ID"].astype(str), sort=False):
        g = g.sort_values("time")
        first = g.iloc[0]
        lost = g.loc[g["undrogued"], "time"]
        s = splits.get(bid, [])
        rows.append({
            "id": bid,
            "start": _iso(first["time"]),
            "end": _iso(g["time"].iloc[-1]),
            "lat0": round(float(first["lat"]), COORD_DP),
            "lon0": round(float(to_display_longitude(np.array([first["lon"]]))[0]), COORD_DP),
            "fixes": int(len(g)),
            "products": sorted(set(g["product"])) if "product" in g.columns else ["hourly"],
            "tier": _tier_summary(g),
            "drogue_lost": _iso(lost.iloc[0]) if len(lost) and not lost.index.equals(g.index) else None,
            "splits": s,
            "sealed": bool({"sealed", "holdout-2023"} & set(s)),
            "file": f"{TRACK_DIR}/{bid}.json",
        })
    return sorted(rows, key=lambda r: (r["start"], r["id"]))


def publish_tracks(df: pd.DataFrame, units: pd.DataFrame | None, out: Path) -> dict:
    """Write the index and every buoy's track file under `out`; return what was written."""
    tracks = out / TRACK_DIR
    tracks.mkdir(parents=True, exist_ok=True)
    index = buoy_index(df, units)
    sizes = []
    for bid, g in df.groupby(df["ID"].astype(str)):
        path = tracks / f"{bid}.json"
        path.write_text(json.dumps(track_record(g), separators=(",", ":")), encoding="utf-8")
        sizes.append(path.stat().st_size)
    manifest = {
        "generated": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "product": "drifter_tracks",
        "label": "NOAA GDP drifters, one file per buoy",
        "longitude_convention": "-180..180 (display; D020 stores 0-360)",
        "buoys": index,
        "counts": {"buoys": len(index), "fixes": int(len(df)),
                   "sealed_buoys": sum(r["sealed"] for r in index)},
        "provenance": {"script": "sar.viz.drifters --tracks",
                       "split": "sar.validate.split (vault D025)"},
    }
    index_path = out / INDEX_NAME
    index_path.write_text(json.dumps(manifest, separators=(",", ":")), encoding="utf-8")
    summary = {
        "buoys": len(index),
        "index_bytes": index_path.stat().st_size,
        "track_bytes_total": int(sum(sizes)),
        "track_bytes_median": int(np.median(sizes)) if sizes else 0,
        "track_bytes_max": int(max(sizes)) if sizes else 0,
    }
    print(f"wrote     {INDEX_NAME}  {summary['buoys']} buoys, {summary['index_bytes'] / 1e3:.0f} kB")
    print(f"wrote     {TRACK_DIR}/  {summary['track_bytes_total'] / 1e6:.1f} MB in total; "
          f"median {summary['track_bytes_median'] / 1e3:.0f} kB, "
          f"largest {summary['track_bytes_max'] / 1e3:.0f} kB per buoy")
    return summary


def main() -> None:
    p = argparse.ArgumentParser(description="Publish GDP drifters for the site.")
    p.add_argument("--data", required=True, help="archive root holding raw/gdp_*.csv")
    p.add_argument("--out", required=True, help="where the published files go. No default.")
    p.add_argument("--source", help="which gdp_hourly_*.csv, if the archive holds several")
    p.add_argument("--tracks", action="store_true",
                   help="write the per-buoy track export (#50) instead of the Parquet: "
                        "both products, with the validation split from derived/validation_split.csv")
    args = p.parse_args()
    if args.tracks:
        # Imported here so the Parquet path keeps working on a checkout without it.
        from sar.validate.split import load_study_drifters

        split_csv = Path(args.data) / "derived" / "validation_split.csv"
        if not split_csv.exists():
            raise SystemExit(f"no {split_csv}; run `python -m sar.validate.split --data {args.data}` first")
        publish_tracks(load_study_drifters(args.data), pd.read_csv(split_csv), Path(args.out))
        return
    publish(Path(args.data), Path(args.out),
            Path(args.source) if args.source else None)


if __name__ == "__main__":
    main()

File: frontend/src/drifters.js
/**
 * drifters.js -- the logic behind the drifter layer, with no map and no DOM (issue #50).
 *
 * The drifters are the project's TEST DATA: real buoys the drift engine is scored
 * against (vault D018, D025). The layer does two jobs, kept apart on purpose:
 *
 *   FINDING   follows the span control. Every buoy seen in the current window is a
 *             faint dot where it was first seen; the list and its search hold the same
 *             buoys, and the search reaches every buoy whatever the window.
 *   WATCHING  follows the clock. Buoys in the water at the current moment are solid,
 *             at their position now, and move on play. The selected buoy's path draws
 *             behind it as time passes.
 *
 * Everything here is a plain function of plain data, so it is tested without a
 * browser, the way geo.js and drift.js are.
 */

const HOUR_MS = 3600 * 1000;

/**
 * A track file as typed arrays: times in epoch ms, and per-fix flags.
 *
 * The file stores whole hours since `t0`, which is exact for both GDP products
 * (every fix is on the hour). Lengths must agree, or the file is refused by name.
 */
export function decodeTrack(json) {
  const n = json.h.length;
  for (const key of ['lat', 'lon', 'seg', 'und', 'far']) {
    if (json[key].length !== n) {
      throw new Error(`track ${json.id}: ${key} has ${json[key].length} values for ${n} fixes`);
    }
  }
  const t0 = Date.parse(json.t0);
  const t = new Float64Array(n);
  for (let k = 0; k < n; k += 1) t[k] = t0 + json.h[k] * HOUR_MS;
  return {
    id: json.id,
    t,
    lat: Float64Array.from(json.lat),
    lon: Float64Array.from(json.lon),
    seg: Int32Array.from(json.seg),
    und: Uint8Array.from(json.und),
    far: Uint8Array.from(json.far),
  };
}

/** The index's buoys with their times parsed once, sorted by first fix. */
export function prepareIndex(manifest) {
  return manifest.buoys
    .map((b) => ({ ...b, startMs: Date.parse(b.start), endMs: Date.parse(b.end) }))
    .sort((a, b) => a.startMs - b.startMs || (a.id < b.id ? -1 : 1));
}

/** Buoys whose record overlaps the window `[winStart, winEnd)`, in ms. */
export function inWindow(entries, winStart, winEnd) {
  return entries.filter((e) => e.startMs < winEnd && e.endMs >= winStart);
}

/**
 * Buoys whose record spans the moment `when`. A buoy can still be in a gap between
 * segments at that moment; `positionAt` says so by returning null.
 */
export function spanningNow(entries, when) {
  return entries.filter((e) => e.startMs <= when && e.endMs >= when);
}

const DATE_QUERY = /^\d{4}(-\d{2}(-\d{2})?)?$/;

/** The UTC period a date query names, as `[start, end)` in ms. */
function periodOf(query) {
  const [y, m, d] = query.split('-').map(Number);
  if (d) return [Date.UTC(y, m - 1, d), Date.UTC(y, m - 1, d + 1)];
  if (m) return [Date.UTC(y, m - 1, 1), Date.UTC(y, m, 1)];
  return [Date.UTC(y, 0, 1), Date.UTC(y + 1, 0, 1)];
}

/**
 * Search EVERY buoy, whatever the window: this is the one way out of the span filter,
 * so a buoy named in the paper can always be found.
 *
 * Matches an ID fragment, or a date (`2021`, `2021-03`, `2021-03-05`) that the buoy's
 * record overlaps. A bare year could be either, so it matches both.
 *
 * For a date, buoys FIRST SEEN in that period come first. A long-lived buoy that only
 * passes through it still matches, but listed first it sent a search for March 2021 to
 * a buoy first seen on 1 January 2019, which is correct and not what anyone meant.
 */
export function searchBuoys(entries, query) {
  const q = (query ?? '').trim();
  if (!q) return [];
  const byDate = DATE_QUERY.test(q) ? periodOf(q) : null;
  const hits = entries.filter((e) => e.id.includes(q)
    || (byDate && e.startMs < byDate[1] && e.endMs >= byDate[0]));
  if (!byDate) return hits;
  const startsIn = (e) => e.startMs >= byDate[0] && e.startMs < byDate[1];
  return hits.sort((a, b) => (startsIn(b) - startsIn(a)) || (a.startMs - b.startMs));
}

/**
 * The smallest span whose window holds the buoy's whole record, or null for the
 * whole archive.
 *
 * The clock aligns a window to a whole multiple of its span from the epoch (a day is a
 * UTC day), so "covers the track" means the ALIGNED window containing the first fix
 * also contains the last one. A track that crosses a boundary needs the next span up.
 */
export function spanCovering(startMs, endMs, spansSeconds) {
  const spans = spansSeconds.filter((s) => s > 0).sort((a, b) => a - b);
  for (const s of spans) {
    const ms = s * 1000;
    const from = Math.floor(startMs / ms) * ms;
    if (endMs < from + ms) return s;
  }
  return null;
}

/** Index of the last fix at or before `when`, or -1 if `when` is before the first. */
export function lastFixAtOrBefore(track, when) {
  const { t } = track;
  let lo = 0;
  let hi = t.length - 1;
  if (t.length === 0 || when < t[0]) return -1;
  while (lo < hi) {
    const mid = (lo + hi + 1) >> 1;
    if (t[mid] <= when) lo = mid; else hi = mid - 1;
  }
  return lo;
}

/**
 * Where the buoy is at `when`: on a fix, or interpolated between two fixes of the SAME
 * segment. Null before the first fix, after the last, and inside a gap between
 * segments, because a buoy with no fixes for days is not somewhere we can draw.
 */
export function positionAt(track, when) {
  const k = lastFixAtOrBefore(track, when);
  if (k < 0) return null;
  if (track.t[k] === when) return [track.lat[k], track.lon[k]];
  if (k === track.t.length - 1 || track.seg[k] !== track.seg[k + 1]) return null;
  const w = (when - track.t[k]) / (track.t[k + 1] - track.t[k]);
  return [
    track.lat[k] + w * (track.lat[k + 1] - track.lat[k]),
    track.lon[k] + w * (track.lon[k + 1] - track.lon[k]),
  ];
}

/** The drogue state at `when`: that of the last fix at or before it, or null. */
export function undroguedAt(track, when) {
  const k = lastFixAtOrBefore(track, when);
  return k < 0 ? null : track.und[k] === 1;
}

/**
 * The path drawn so far, up to `when`, as pieces with one style each.
 *
 * Each step between two consecutive fixes of one segment gets the drogue state of
 * the fix it arrives at, and is `far` if either end is more than 3 h from a real
 * GPS fix. Consecutive steps with the same style join into one piece. No step is
 * drawn across a gap between segments. The last piece ends at the interpolated
 * position now, so the trail stays attached to the marker between fixes.
 */
export function trailUpTo(track, when) {
  const pieces = [];
  const last = lastFixAtOrBefore(track, when);
  let current = null;
  for (let k = 1; k <= last; k += 1) {
    if (track.seg[k] !== track.seg[k - 1]) { current = null; continue; }
    const undrogued = track.und[k] === 1;
    const far = track.far[k] === 1 || track.far[k - 1] === 1;
    if (!current || current.undrogued !== undrogued || current.far !== far) {
      current = { undrogued, far, points: [[track.lat[k - 1], track.lon[k - 1]]] };
      pieces.push(current);
    }
    current.points.push([track.lat[k], track.lon[k]]);
  }
  const now = positionAt(track, when);
  if (now && last >= 0 && track.t[last] !== when) {
    const undrogued = track.und[last] === 1;
    const far = track.far[last] === 1 || track.far[last + 1] === 1;
    if (!current || current.undrogued !== undrogued || current.far !== far) {
      current = { undrogued, far, points: [[track.lat[last], track.lon[last]]] };
      pieces.push(current);
    }
    current.points.push(now);
  }
  return pieces;
}

/**
 * Where the selected buoy's record sits on the slider, as fractions of the window:
 * `{ left, width }`, clipped to the window, or null if it does not overlap.
 */
export function lifetimeBar(startMs, endMs, winStart, winEnd) {
  const span = winEnd - winStart;
  if (!(span > 0) || endMs < winStart || startMs >= winEnd) return null;
  const left = Math.max(0, (startMs - winStart) / span);
  const right = Math.min(1, (endMs - winStart) / span);
  return { left, width: Math.max(right - left, 0.004) };
}

const TIER_WORDS = {
  drogued: 'drogued: follows the water',
  undrogued: 'undrogued: the wind pushes it too',
  mixed: 'loses its drogue part-way',
  uncertain: 'drogue records disagree',
};

/** `2019-03-02`: a date as a reader names it. Every source here is UTC. */
export function day(ms) {
  return new Date(ms).toISOString().slice(0, 10);
}

/** Whole days a record spans, at least one. */
export function lengthDays(entry) {
  return Math.max(1, Math.round((entry.endMs - entry.startMs) / (24 * HOUR_MS)));
}

/** The words a tooltip or a list row shows for one buoy. */
export function describeBuoy(entry) {
  const days = lengthDays(entry);
  return {
    title: `Buoy ${entry.id}`,
    dates: `${day(entry.startMs)} → ${day(entry.endMs)} · ${days} day${days === 1 ? '' : 's'}`,
    tier: TIER_WORDS[entry.tier] ?? entry.tier,
    sealed: entry.sealed
      ? 'sealed: look at the track, but do not debug the engine on it (D025)'
      : null,
  };
}
